README.md

```diff
@@ -0,0 +1,230 @@
+
+```sh
+> qemu-guest -k elfloader_kvm-x86_64 -e rootfs/ \
+            -a "env.vars=[ LD_LIBRARY_PATH=/lib LD_SHOW_AUXV=1 ] -- /helloworld <application arguments>"
+```
+
+## Debugging
+
+### `strace`-like Output
+
+Unikraft's [`syscall_shim`](https://github.com/unikraft/unikraft/tree/staging/lib/syscall_shim) provides the ability to print a strace-like message for every processed binary system call request on the kernel output. This option can be useful for understanding what code a system call handler returns to the application and how the application interacts with the kernel.
+The setting can be found under `Library Configuration -> syscall_shim -> Debugging`: `'strace'-like messages for binary system calls`.
+
+### GNU Debugger (gdb)
+
+It is possible to debug `elfloader` together with the loaded application and use the full set of debugging facilities for kernel and application at the same time.
+In principle, `gdb` must only be made aware of the runtime memory layout of `elfloader` with the loaded application.
+Thanks to the single address space layout, we gain easy debugability and full transparency.
+
+#### Static-PIE ELF executable
+
+As a first step, `gdb` is started with loading the symbols from the `dbg` image of the `elfloader`. We map the symbols of ELF application with the gdb command `add-symbol-file` by specifying the application (with debug symbols) and the base load address.
+If `info` messages are enabled in `ukdebug`, this base address will be messaged by the loader like this:
+
+```
+ELF program loaded to 0x400101000-0x4001c2a08 (793096 B), entry at 0x40010ad50
+```
+
+To this address (here: `0x400101000`) you have to add the offset of the `.text` segment.
+You can use `readelf -S` to find it out. In our example it is `0x92a0` (output shortened):
+
+```
+$ readelf -S helloworld_static
+Section Headers:
+  [Nr] Name              Type             Address           Offset
+       Size              EntSize          Flags  Link  Info  Align
+  [11] .text             PROGBITS         00000000000092a0  000092a0
+       0000000000086eb0  0000000000000000  AX       0     0     32
+```
+
+The resulting address here is `0x40010a2a0`. The symbols of the static helloworld program can then be loaded from `gdb` with the following command:
+
+```
+(gdb) add-symbol-file -readnow helloworld_static 0x40010A2A0
+```
+
+From this point you have symbol resolution in your debugger for both, the Unikraft elfloader and the loaded application.
+
+NOTE: You can only set regular breakpoints within the application (`break` with GDB) after it got loaded into memory by elfloader (otherwise they will be ignored).
+The recommended procedure is 1) to set a breakpoint just after the application was loaded (e.g., the first system call that the application executes), 2) let the execution continue until the breakpoint is reached, and 3) set the interesting breakpoints within application space.
+
+#### Dynamically linked ELF executable
```

Comment:
```suggestion
#### Dynamically-linked ELF
```

Reply:
:+1:

Reply:
I'll keep executable because we kept it also for `Static-PIE ELF executable`